### PR TITLE
Add rfc doc for default idp configurable

### PR DESCRIPTION
**Pull request by dengmingtong**
[rendered html version](https://github.com/dengmingtong/service-workbench-on-aws-cn/blob/mingtong_cn_rfc/rfcs/default_idp_configurable/default_idp_configurable.md)

Issue #, if available:

RFC for #12 

Description of changes:

Checklist:

- [x] Have you successfully deployed to an AWS account with your changes?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully tested with your changes locally?
- [ ] If new dependencies have been added, have they been pinned to specific versions?
- [ ] Is this change also required on the AWS Solution version?
- [ ] Have you updated openapi.yaml if you made updates to API definition (including add, delete or update parameter and request data schema)?
- [ ] If you had to run manual tests, have you considered automating those tests by adding them to [end-to-end tests](../main/end-to-end-tests/README.md)?
- [ ] If you are updating the changelog and vending out a new release, have you updated versionNumber and versionDate in [.defaults.yml](../main/config/settings/.defaults.yml)

AS review ticket id:

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: rfcs/default_idp_configurable/default_idp_configurable.md [hunk 1]
```diff
@@ -0,0 +1,393 @@
+---
+feature name: default-idp-configurable
+start date: May 9th, 2022
+Summary: Make the default IDP of Service Workbench solution configurable
+rfc pr: 
+related issue: https://github.com/awslabs/service-workbench-on-aws-cn/issues/12
+---
+
+# Summary
+This RFC proposes to make the Service Workbench's default IDP configurable in order to allow user to choose Keycloak as default IDP in the regions without [Amazon Cognito user pools][cognito-user-pool].
```
Comment (zxkane) [inline rfcs/default_idp_configurable/default_idp_configurable.md]: the goal is supporting other IdPs besides Cognito, such as KeyCloak, Okta, Authing and etc.

File: rfcs/default_idp_configurable/default_idp_configurable.md [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+
+# Motivation
+Service Workbench is an open-source web application that provides a self-service and push-button provisioning experience of ready-to-use IT environments for research projects or classroom experiments on AWS. The authentication and authorization are mandatory features. User must be registered and granted before they can use Service Workbench web console. 
+
+## What are we doing today?
+The current situation is that Service Workbench already includes authentication and authorization features. The Service Workbench use Cognito User Pool as the default IDP to implement authentication feature. The Service Workbench also support users to use other IDPs as federated IDPs to implementation authentication feature.
```
Comment (zxkane) [inline rfcs/default_idp_configurable/default_idp_configurable.md]: > The Service Workbench also support users to use other IDPs as federated IDPs to implementation authentication feature.

Do you mean other federated IdPs are provided by Cognito?

File: rfcs/default_idp_configurable/default_idp_configurable.md [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+The current situation is that Service Workbench already includes authentication and authorization features. The Service Workbench use Cognito User Pool as the default IDP to implement authentication feature. The Service Workbench also support users to use other IDPs as federated IDPs to implementation authentication feature. 
+The authorization feature is implemented by Service Workbench internal logic, the user's permission is managed in Service Workbench DynamoDB.
+
+For authentication feature, either we use Cognito User Pool as IDP or use other IDPs as federated IDPs, we have to use Cognito User Pool.
+This implementation causes two issues:
+1. Cognito user pool become a blocking issue of deploying Service Workbench in the regions without Cognito User Pool.
+2. Users will concern that they have to use Cognito User Pool. Some users would like to use other IDP directly, such as Keycloak.
+
+## What are we proposing to do?
+This RFC proposes to release the Service Workben to allow user to choose Keycloak to implement authentication feature without Cognito User Pool.
```
Comment (zxkane) [inline rfcs/default_idp_configurable/default_idp_configurable.md]: The goal of this RFC is support third party IdPs for authentication in Service Workbench, KeyCloak will be a built-in implementation in this change.

File: rfcs/default_idp_configurable/default_idp_configurable.md [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+API Gateway handles all private request and verify if the token is valid through the lambda function(AuthenticationLayerHandler). The lambda function(AuthenticationLayerHandler) get "tokenValidatorLocator" from "Auth Provider Config", then invoke validateToken method to verify input token.
+
+If user choose federated IDP to authenticate user, the main sequence is the same with Cognito User Pool, there is only a little different for user logining and getting id_token. The callflow like below:
+
+![amazon-cognito-user-pools-saml-idp-auth-flow](./amazon-cognito-user-pools-saml-idp-auth-flow.png)
+
+The user is also redirected to Cognito User Pool Server, but the URL contains **idp_identifier=idpId", Cognito User Pool will go on redirecting to IdP.
+
+# Design Summary
+The general approach proposed by this RFC is to add some configurations to allow Service Workbench to integrate with Keycloak. This will make Service Workbench available in the regions where Cognito User Pool is not available.
```
Comment (zxkane) [inline rfcs/default_idp_configurable/default_idp_configurable.md]: The general approach proposed by this RFC is supporting third-party IdPs for authentication. Below proposed implementation uses KeyCloak as a built-in third-party IdP for reference.

File: rfcs/default_idp_configurable/default_idp_configurable.md [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+
+## Detail Design
```
Comment (zxkane) [inline rfcs/default_idp_configurable/default_idp_configurable.md:139]: I propose the first part demonstrates the high level changes for supporting third-party IdP.

Then go to the detail how implementing the third-party IdP, such as KeyCloak.

File: rfcs/default_idp_configurable/default_idp_configurable.md [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+2. Create a realm for Service Workbench or use existed realm.
+3. Create a client for Service Workbench.
+4. Create root user in Keycloak.
+
+We will add following configurations options for keycloak integration.
+```
+keyCloakRealm: xxx
+keyCloakClientId: xxx
+keyCloakUrl: xxx
+rootUserEmail: xxx
```
Comment (zxkane) [inline rfcs/default_idp_configurable/default_idp_configurable.md]: Are below options supported already in Cognito? If not, is there any existing options to be shared?

```
rootUserEmail: xxx
rootUserFirstName: xxx
rootUserLastName: xxx
```

File: rfcs/default_idp_configurable/default_idp_configurable.md [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+rootUserFirstName: xxx
+rootUserLastName: xxx
+```
+![deployment_keycloak](./swb_auth_deploy_keycloak.drawio.png)
```
Comment (zxkane) [inline rfcs/default_idp_configurable/default_idp_configurable.md:152]: I think the `AddAuthProviders` has two implementations, using the expected one depends on the IdP type configuration. Should polish the architect diagram to reflect the implementation.

File: rfcs/default_idp_configurable/default_idp_configurable.md [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+    ...
+    "tokenValidatorLocator": "locator:service:keycloakAuthenticationProviderService/validateToken",
+    "tokenRevokerLocator": "locator:service:keycloakAuthenticationProviderService/revokeToken"
+    ...
+  }
+  ...
+}
+```
+
+#### KeycloakAuthenticationProviderService
```
Comment (zxkane) [inline rfcs/default_idp_configurable/default_idp_configurable.md]: Is it a runtime service for authentication?

File: rfcs/default_idp_configurable/default_idp_configurable.md [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+        break;
+      default:
+        console.log(`Sorry, we are out of ${defaultIDPType}.`);
+    }    
+  }
+``` 
+
+The difference with createCognitoNativeAdminUser(), createKeycloakNativeAdminUser does not need to create user in Keycloak, just need to create root user in Service Workbench DynamoDB, the user in Keycloak should be created before deploying Service Workbench. We add some configuration items in main/config/settings/${STAGE_NAME}.yml to configure root user information.
+```
+rootUserEmail: xxx
```
Comment (zxkane) [inline rfcs/default_idp_configurable/default_idp_configurable.md:242]: Looks like those user configurations are only for KeyCloak, add prefix `keyCloak` is better

File: rfcs/default_idp_configurable/default_idp_configurable.md (hunk 1 above)
Comment (dengmingtong) [reply rfcs/default_idp_configurable/default_idp_configurable.md]: Done

File: rfcs/default_idp_configurable/default_idp_configurable.md (hunk 2 above)
Comment (dengmingtong) [reply rfcs/default_idp_configurable/default_idp_configurable.md]: Yes, user can use other federated IdPs through Cognito.

File: rfcs/default_idp_configurable/default_idp_configurable.md (hunk 3 above)
Comment (dengmingtong) [reply rfcs/default_idp_configurable/default_idp_configurable.md]: OK

File: rfcs/default_idp_configurable/default_idp_configurable.md (hunk 4 above)
Comment (dengmingtong) [reply rfcs/default_idp_configurable/default_idp_configurable.md]: OK

File: rfcs/default_idp_configurable/default_idp_configurable.md (hunk 5 above)
Comment (dengmingtong) [reply rfcs/default_idp_configurable/default_idp_configurable.md:139]: Good point!

File: rfcs/default_idp_configurable/default_idp_configurable.md (hunk 6 above)
Comment (dengmingtong) [reply rfcs/default_idp_configurable/default_idp_configurable.md]: Yes,  these configurations are used for Cognito.  For cognito idp, the root user will be created in cognito and Service Workbench DynamoDB based on these configurations during the deployment process. For keycloak idp, we need to create root user in keycloak before deploying Service Workbench.  So during the deployment process, these configurations are used for creating root user into Service Workbench DynamoDB

File: rfcs/default_idp_configurable/default_idp_configurable.md (hunk 9 above)
Comment (dengmingtong) [reply rfcs/default_idp_configurable/default_idp_configurable.md:242]: As I explained in previous conversation, these configurations are used for both keycloak and cognito. My doc description caused a misunderstanding, I will refactor it.

File: rfcs/default_idp_configurable/default_idp_configurable.md (hunk 7 above)
Comment (dengmingtong) [reply rfcs/default_idp_configurable/default_idp_configurable.md:152]: Yes, I agree, I will refactor the doc.

File: rfcs/default_idp_configurable/default_idp_configurable.md (hunk 8 above)
Comment (dengmingtong) [reply rfcs/default_idp_configurable/default_idp_configurable.md]: Yes, it is runtime service for authenctication, I will move it to runtime part.

File: rfcs/default_idp_configurable/default_idp_configurable.md [hunk 10]
```diff
@@ -0,0 +1,435 @@
...
+
+## What are we proposing to do?
+This RFC proposes to release the Service Workben to support third party IdPs for authentication in Service Workbench, KeyCloak will be a built-in implementation in this change.
+
+After releasing, the Service Workbench should:
+1. Support users to deploy and use Service Workbench solution in the regions without Cognito User Pool.
+2. Support the case that the users don't want to use Cognito User Pool, but want to use Keycloak as default IDP. 
+3. Make the Service Workbench has well architecture to extend other IdPs which support OpenId-Connect.
+
+# Authentication of Cognito User Pool
```
Comment (zxkane) [inline rfcs/default_idp_configurable/default_idp_configurable.md]: might be 'Current authentication implementation via Cognito User Pool'

File: rfcs/default_idp_configurable/default_idp_configurable.md [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+
+The user is also redirected to Cognito User Pool Server, but the URL contains **idp_identifier=idpId", Cognito User Pool will go on redirecting to federated IdP.
+
+# Design Summary
+The general approach proposed by this RFC is supporting other third-party IdPs for authentication beside Cognito User Pool. Below proposed implementation uses KeyCloak as a built-in third-party IdP for reference.
+
+## High level design.
+The proposal of high level design as below:
+
+![swb_auth_design_summary](./swb_auth_design_summary.drawio.png)
```
Comment (zxkane) [inline rfcs/default_idp_configurable/default_idp_configurable.md:135]: New high level of arch diagram is good. It'd better to create placeholders for 

- identity providers, which would be `Cognito User Pool`, `KeyCload`, `Okta` and other implementations
- create root user in IdP
- runtime functions to verify the tokens against the implementation

File: rfcs/default_idp_configurable/default_idp_configurable.md [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+
+We will refactor green parts: **AddAuthProviders**, **CreateRootUser** and **implementation of JWT token verification**, make these parts support third-party IdPs based on configuration.
+
+## Detail Design
+The design will be divided into two parts: Deployment Process and Runtime Process.
+
+### Deployment Process design
+We keep the current highlevel design that invoking lambda function(postDeployment) to create the resources related with the authentication feature. But which resources are created is determined by configuration as below in "main/config/settings/${STAGE_NAME}.yml" file:
+```
+defaultIDPType: keycloak/cognito
```
Comment (zxkane) [inline rfcs/default_idp_configurable/default_idp_configurable.md]: which IdP type is the default one if this field is missing?

File: rfcs/default_idp_configurable/default_idp_configurable.md (hunk 10 above)
Comment (dengmingtong) [reply rfcs/default_idp_configurable/default_idp_configurable.md]: Fixed

File: rfcs/default_idp_configurable/default_idp_configurable.md (hunk 11 above)
Comment (dengmingtong) [reply rfcs/default_idp_configurable/default_idp_configurable.md:135]: Fixed

File: rfcs/default_idp_configurable/default_idp_configurable.md (hunk 12 above)
Comment (dengmingtong) [reply rfcs/default_idp_configurable/default_idp_configurable.md]: The default value will be added into main/config/settings/.defaults.yml files. We set the default value as keycloak for service-workbench-on-aws-cn repo. I will refactor rfc doc to describe it clearly.

File: rfcs/default_idp_configurable/default_idp_configurable.md [hunk 13]
```diff
@@ -0,0 +1,435 @@
...
+# Current authentication implementation via Cognito User Pool
+
+## High level design for Cognito User Pool authentication
+
+![swb_auth_cognito_summary](./swb_auth_cognito_summary.drawio.png)
+
+According to this diagram, we can know that authentication related two parts: deployment process and runtime process.
+
+## Deployment process
+During Service Workbench deployment, the authentication related resources will be created in lambda function(postDeployment), such as creating Cognito User Pool, generating **Auth Provider Config** and saving it into  DynamoDB, create root user and so on.
```
Comment (chenhaiyun) [inline rfcs/default_idp_configurable/default_idp_configurable.md:40]: 如果是 其他的 IDP，这个创建 root 用户创建的逻辑是什么？如何创建？

File: rfcs/default_idp_configurable/default_idp_configurable.md [hunk 14]
```diff
@@ -0,0 +1,435 @@
...
+    ...
+  }
+]
+```
+
+- Get id_token. 
+The Service Workbench use **response_type=code** way to integrate with Cognito User Pool, browser redirect to Cognito User Pool login page according to above **Providers Config**. After user input correct username and password, browser redirect user back to Service Workbench page, then request id_token from Service Workbench server-side lambda function(ApiHandler). The lambda function(ApiHandler) send request withing **code** to Cognito user Pool server for id_token.
+
+- Verify id_token.
+API Gateway handles all private request and verify if the token is valid through the lambda function(AuthenticationLayerHandler). The lambda function(AuthenticationLayerHandler) get "tokenValidatorLocator" from "Auth Provider Config", then invoke validateToken method to verify input token.
```
Comment (chenhaiyun) [inline rfcs/default_idp_configurable/default_idp_configurable.md:121]: 这个验证 id token 的 API gateway 是我们的解决方案需要实现的？

File: rfcs/default_idp_configurable/default_idp_configurable.md [hunk 15]
```diff
@@ -0,0 +1,435 @@
...
+The current situation is that Service Workbench already includes authentication and authorization features. The Service Workbench use Cognito User Pool as the default IdP to implement authentication feature. The Service Workbench also support users to use other IdPs as federated IdP through Cognito User Pool. 
+The authorization feature is implemented by Service Workbench internal logic, the user's permission is managed in Service Workbench DynamoDB.
+
+For authentication feature, either we use Cognito User Pool as IdP or use other IdPs as federated IdPs, we have to use Cognito User Pool.
+This implementation causes two issues:
+1. Cognito user pool become a blocking issue of deploying Service Workbench in the regions without Cognito User Pool.
+2. Users will concern that they have to use Cognito User Pool. Some users would like to use other IdP directly, such as Keycloak.
+
+## What are we proposing to do?
+This RFC proposes to release the Service Workben to support third party IdPs for authentication in Service Workbench, KeyCloak will be a built-in implementation in this change.
```
Comment (chenhaiyun) [inline rfcs/default_idp_configurable/default_idp_configurable.md]: 是不是部署的时候，需要2套CloudFormation的模版？如果用户不想用KeyCloak，那就不用帮用户创建对应的资源吧？

**Review by wchaws — COMMENTED** [review]
rfcs/default_idp_configurable/swb_auth_runtime_cognito.drawio.png
1. 这个图里面为什么是拿 Id_Token 去访问 Private API，不应该是 Access Token 么？
2. Id token 为什么要去 backend 绕一圈拿，而不是直接通过 WebUI 去 Cognito User Pool 拿？

File: rfcs/default_idp_configurable/default_idp_configurable.md [hunk 16]
```diff
@@ -0,0 +1,435 @@
...
+AddAuthProviders class is responsible for creating cognito related resources and generate **Auth Provider Config** and save it into  DynamoDB. If there are federated IdP configs, AddAuthProviders is also responsible for creating "Identity providers" in Cognito User Pool, and save federated IdP config into DynamoDB.
+
+**Cognito related resources includes:**
+  - Cognito User Pool
+  - User Pool Client
+  - User Pool Domain
+  - Identity Provider
+  - Attribute Mapping
+
+The **Auth Provider Config** is the key configuration for each IdP, and it is used by both the client-side and back-end, mainly contains the following three parts:
```
Comment (wchaws) [inline rfcs/default_idp_configurable/default_idp_configurable.md:54]: 这个配置文件能通过 `.well-known/openid-configuration` endpoint 自动去拿么？
大部分的 IDP 应该支持 `.well-known/openid-configuration`

File: rfcs/default_idp_configurable/default_idp_configurable.md (hunk 13 above)
Comment (dengmingtong) [reply rfcs/default_idp_configurable/default_idp_configurable.md:40]: 如果是其他IDP，创建root用户的逻辑：需要用户先在IDP里面创建root用户，然后将root用户的相关信息设定到配置文件中。在文档的后面有说明

File: rfcs/default_idp_configurable/default_idp_configurable.md (hunk 14 above)
Comment (dengmingtong) [reply rfcs/default_idp_configurable/default_idp_configurable.md:121]: 是的，API Gateway会集成一个解决方案实现的lambda function

File: rfcs/default_idp_configurable/default_idp_configurable.md (hunk 15 above)
Comment (dengmingtong) [reply rfcs/default_idp_configurable/default_idp_configurable.md]: 一套cloudformation 模版，会通过配置文件来创建不同的资源

File: rfcs/default_idp_configurable/default_idp_configurable.md (hunk 16 above)
Comment (dengmingtong) [reply rfcs/default_idp_configurable/default_idp_configurable.md:54]: good point, well refactor rfc for this.

**Review by zxkane — APPROVED** [review]

**Pull request merged**